### ISSUE TITLE
vdisk created on vpool with unsatisfiable policy is present in model but not in reality

### ISSUE THREAD

**Issue by JeffreyDevloo** [openvstorage/framework#1091]
### TLDR

A minified version of the ticket.

Vdisk fails when created on a vpool that has unsatisfied preset.

Two options to fix it:
- Do not let the vpool be made with the unsatisfied policy

or 
- Report the correct problem back to the user (instead of the 'Failed to create'
- Remove the Vdisk from the DAL when the issue occurred.
### Problem description

When a vdisk has been created through GUI after that the vpool, on which vdisk will be set, was created with an unsatisfiable policy, the vdisk will not be present in reality (not present on filesystem) because the volumedriver will log an unsatisfiable albaproxy policy.
However the vdisk is created in the DAL (without a volume_id as volumedriver couldn't register the object)

The GUI will also log an error during creation of the vdisk but not specifically what it was. (Just the normal 'Failed to create ..'.

How to reproduce:
- Setup environment
- Add roles to disks
- Add backend 
- Initialize 1 disk with 1 asd (default preset will be unsatisfied)
- Create vpool (you will see the warning on the wizard)
- Add a vdisk to the created vpool

If the policy is satisfied afterwards, the vdisk will still be broken (volumedriver can't register an unknown object again). 
### Temporary fix

Remove the vdisk from the DAL using VDisk().remove.
### Possible solution

Two options:
**Fix it by only tackling the issue**
Two step solution:
- Catch the error and output it back to the customer in the GUI
- Catch the error in code and remove the created DAL object when it occured

**Fix it by only tackling the core of the issue**
- Do not let a vpool be created with a unsatisfied preset
### Additional information

This issue could be prevented (in GUI) by blocking the vpool wizard when the policy is unsatisfied.
In my eyes it makes hardly any sense that you would let the customer create a 'faulty' vpool even if it would be temporary.
Currently this behavior is by design and was requested in http://jira.cloudfounders.com/browse/OVS-4466 but we should re-evaluate the benefit of this 'feature'.

If the vpool can be created with an unsatisfiable policy after this issue has been fixed with the deletion of the DAL object, the following must be tested:
- test the impact of the unsatisfiable policy when truncating a volume in the mount point

**Comment by khenderick** [comment]
@wimpers, @saelbrec, do we still want to allow customers to create a vPool when we know the selected preset will cause issues?

**Comment by wimpers** [comment]
It is not because the policy can be satisfied when creating the vPool, that it will also be satisfied when creating the first vDisk. Hence the proposed solution to block the vPool wizard is not a viable fix.

**Comment by wimpers** [comment]
> do we still want to allow customers to create a vPool when we know the selected preset will cause issues?

Yes, as you can always add more disks or the policy can be unstisfiable after the vPool creation. The no satisfiable policy should be tacked when the disk is created.
- In the create disk API, check for a policy and catch the error and cleanup if you would still run into it (f.e. after check, some disks die)
- in the GUI: block the create disk when no satisfiable policy is available for the vPool. (hovver message 'no satisfiable backend policy')
- when a disk is created through FUSE, catch the error and cleanup the vdisk.

**Comment by khenderick** [comment]
@wimpers, that would mean we always have to call to the remote backend-hosting cluster to validate the policies over there. I don't think that will be feasible.

**Comment by wimpers** [comment]
@khenderick that is only in case you use an external backend right? Maybe add an api  call for satisfiable policies?

**Comment by khenderick** [comment]
It is available, but I don't feel comfortable calling to the other end too often. Besides the fact that we still don't cover all scenarios (e.g. edge and fuse).

**Comment by khenderick** [comment]
Fixed by #1102, packaged in openvstorage-2.7.4-rev.4238.98a1f23

What is implemented:
- Customers can always create/clone/... vDisks as the framework does not call the remote backend every time to check whether the policy allows new namespaces to be created, since it's quite some work, might be slow, might be blocking vDisk creation, can't prevent vDisks from being created through edge, fuse, ... so it didn't make much sense to put work in some half-assed solution.
- If the creation/cloning/... fails (for whetever reason), everything should be cleaned up properly.

**Comment by JeffreyDevloo** [comment]
### Steps
- Installed latest cluster
- Added disk roles
- Added backend (default preset)
- Claimed 1 asd on node 
- Added vpool
- Added vdisk

### Output
Found the following error log:

```
Nov 21 09:55:11 ovs-node-2 celery[2572]: 2016-11-21 09:55:11 68000 +0100 - ovs-node-2 - 2572/140121143334656 - celery/celery.worker.job - 104 - ERROR - Task ovs.vdisk.create_new[9f99b6da-31a8-4633-b6f7-d467b1b03923] raised unexpected: RuntimeError('got fault response volumeCreate',)
Nov 21 09:55:11 ovs-node-2 celery[2572]: Traceback (most recent call last):
Nov 21 09:55:11 ovs-node-2 celery[2572]:   File "/usr/lib/python2.7/dist-packages/celery/app/trace.py", line 240, in trace_task
Nov 21 09:55:11 ovs-node-2 celery[2572]:     R = retval = fun(*args, **kwargs)
Nov 21 09:55:11 ovs-node-2 celery[2572]:   File "/usr/lib/python2.7/dist-packages/celery/app/trace.py", line 438, in __protected_call__
Nov 21 09:55:11 ovs-node-2 celery[2572]:     return self.run(*args, **kwargs)
Nov 21 09:55:11 ovs-node-2 celery[2572]:   File "/opt/OpenvStorage/ovs/lib/vdisk.py", line 615, in create_new
Nov 21 09:55:11 ovs-node-2 celery[2572]:     node_id=str(storagedriver.storagedriver_id))
Nov 21 09:55:11 ovs-node-2 celery[2572]: RuntimeError: got fault response volumeCreate
```
And nothing was stored in the DAL
![selection_064](https://cloud.githubusercontent.com/assets/17570109/20476407/1bb3f1a8-afd1-11e6-87f8-36f21fcf310f.png)

Edit: The GUI displays an error:
![selection_065](https://cloud.githubusercontent.com/assets/17570109/20476786/fdf5bdca-afd2-11e6-9859-b1a2cabca0c0.png)


### Test result
Edit: Test passed. The GUI displays an error after around 30 s.
